### ISSUE TITLE
'Invalid operand types for ICmp instruction' on ppc64

### ISSUE THREAD

**Issue by kalev** [ldc-developers/ldc#1904]
On ppc64, ldc 1.1.0 beta4 builds fine when boostrapping with older ldc 0.17.2. When building 1.1.0 beta4 with itself, I get the following error. This is with llvm 3.9 on Fedora rawhide.

Full build log available at https://kojipkgs.fedoraproject.org//work/tasks/338/16670338/build.log (short lived link)

```
[ 11%] Generating src/core/math.o
cd /builddir/build/BUILD/ldc-1.1.0-beta4-src && /builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2 --output-o -c -I/builddir/build/BUILD/ldc-1.1.0-beta4-src/runtime/druntime/src -I/builddir/build/BUILD/ldc-1.1.0-beta4-src/runtime/druntime/src/gc /builddir/build/BUILD/ldc-1.1.0-beta4-src/runtime/druntime/src/core/math.d -of/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/runtime/src/core/math.o -w -relocation-model=pic -O3 -release -disable-invariants -d-version=Shared
#0 0x00003fff7b63a478 llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/lib64/libLLVM-3.9.so+0x84a478)
#1 0x00003fff7b63a958 (/lib64/libLLVM-3.9.so+0x84a958)
#2 0x00003fff7b638124 llvm::sys::RunSignalHandlers() (/lib64/libLLVM-3.9.so+0x848124)
#3 0x00003fff7b638304 (/lib64/libLLVM-3.9.so+0x848304)
#4 0x00003fff7dc004d8 (linux-vdso64.so.1+0x4d8)
#5 0x0000000010060b24 RootObject::compare(RootObject*) (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x10060b24)
#6 0x00000100082e6350
#7 0x00000000102f07d0 bool __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(CaseStatement*, CaseStatement*)>::operator()<CaseStatement**, CaseStatement**>(CaseStatement**, CaseStatement**) /usr/include/c++/6.2.1/bits/predefined_ops.h:123:0
#8 0x00000000102f07d0 void std::__insertion_sort<CaseStatement**, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(CaseStatement*, CaseStatement*)> >(CaseStatement**, CaseStatement**, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(CaseStatement*, CaseStatement*)>) /usr/include/c++/6.2.1/bits/stl_algo.h:1844:0
#9 0x00000000102f3898 void std::__final_insertion_sort<CaseStatement**, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(CaseStatement*, CaseStatement*)> >(CaseStatement**, CaseStatement**, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(CaseStatement*, CaseStatement*)>) /usr/include/c++/6.2.1/bits/stl_algo.h:1887:0
#10 0x00000000102f3898 void std::__sort<CaseStatement**, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(CaseStatement*, CaseStatement*)> >(CaseStatement**, CaseStatement**, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(CaseStatement*, CaseStatement*)>) /usr/include/c++/6.2.1/bits/stl_algo.h:1968:0
#11 0x00000000102f3898 void std::sort<CaseStatement**, bool (*)(CaseStatement*, CaseStatement*)>(CaseStatement**, CaseStatement**, bool (*)(CaseStatement*, CaseStatement*)) /usr/include/c++/6.2.1/bits/stl_algo.h:4739:0
#12 0x00000000102f3898 ToIRVisitor::visit(SwitchStatement*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:881:0
#13 0x000000001008f9c8 SwitchStatement::accept(Visitor*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/ddmd/statement.h:453:0
#14 0x00000000102ed864 ToIRVisitor::visit(CompoundStatement*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:110:0
#15 0x0000000010071808 CompoundStatement::accept(Visitor*) (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x10071808)
#16 0x00000000102ea7b0 ToIRVisitor::visit(ScopeStatement*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:398:0
#17 0x00000000100727f0 ScopeStatement::accept(Visitor*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/ddmd/statement.h:263:0
#18 0x00000000102ea7b0 ToIRVisitor::visit(ScopeStatement*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:398:0
#19 0x00000000100727f0 ScopeStatement::accept(Visitor*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/ddmd/statement.h:263:0
#20 0x00000000102eb0f8 ToIRVisitor::visit(CaseStatement*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:1104:0
#21 0x00000000100903e4 CaseStatement::accept(Visitor*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/ddmd/statement.h:474:0
#22 0x00000000102ed864 ToIRVisitor::visit(CompoundStatement*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:110:0
#23 0x0000000010071808 CompoundStatement::accept(Visitor*) (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x10071808)
#24 0x00000000102ea7b0 ToIRVisitor::visit(ScopeStatement*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:398:0
#25 0x00000000100727f0 ScopeStatement::accept(Visitor*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/ddmd/statement.h:263:0
#26 0x00000000102f27a0 ToIRVisitor::visit(SwitchStatement*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:939:0
#27 0x000000001008f9c8 SwitchStatement::accept(Visitor*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/ddmd/statement.h:453:0
#28 0x00000000102ed864 ToIRVisitor::visit(CompoundStatement*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:110:0
#29 0x0000000010071808 CompoundStatement::accept(Visitor*) (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x10071808)
#30 0x00000000102ed864 ToIRVisitor::visit(CompoundStatement*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:110:0
#31 0x0000000010071808 CompoundStatement::accept(Visitor*) (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x10071808)
#32 0x00000000102edb5c Statement_toIR(Statement*, IRState*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/statements.cpp:1658:0
#33 0x00000000102a63fc DtoDefineFunction(FuncDeclaration*, bool) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/functions.cpp:1067:0
#34 0x000000001035e78c CodegenVisitor::visit(FuncDeclaration*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/declarations.cpp:412:0
#35 0x0000000010156c64 FuncDeclaration::accept(Visitor*) (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x10156c64)
#36 0x000000001035eb60 CodegenVisitor::visit(StructDeclaration*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/declarations.cpp:205:0
#37 0x0000000010214f84 StructDeclaration::accept(Visitor*) (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x10214f84)
#38 0x000000001035dd50 CodegenVisitor::visit(AttribDeclaration*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/declarations.cpp:485:0
#39 0x00000000100486c4 Visitor::visit(ProtDeclaration*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/ddmd/visitor.h:386:0
#40 0x000000001027d288 ProtDeclaration::accept(Visitor*) (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x1027d288)
#41 0x000000001035f808 Declaration_codegen(Dsymbol*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/declarations.cpp:576:0
#42 0x00000000102bf878 codegenModule(IRState*, Module*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/gen/modules.cpp:635:0
#43 0x0000000010372748 ldc::CodeGenerator::emit(Module*) /builddir/build/BUILD/ldc-1.1.0-beta4-src/driver/codegenerator.cpp:234:0
#44 0x0000000010345e2c codegenModules(Array<Module*>&) /builddir/build/BUILD/ldc-1.1.0-beta4-src/driver/main.cpp:1127:0
#45 0x0000000010131e40 mars_mainBody(Array<char const*>&, Array<char const*>&) (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x10131e40)
#46 0x0000000010344058 cppmain(int, char**) /builddir/build/BUILD/ldc-1.1.0-beta4-src/driver/main.cpp:1101:0
#47 0x0000000010040b54 _Dmain (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x10040b54)
#48 0x00003fff7a624f78 _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv (/lib64/libdruntime-ldc.so.71+0xe4f78)
#49 0x00003fff7a624dec _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ7tryExecMFMDFZvZv (/lib64/libdruntime-ldc.so.71+0xe4dec)
#50 0x00003fff7a624c5c _d_run_main (/lib64/libdruntime-ldc.so.71+0xe4c5c)
#51 0x0000000010040c7c main (/builddir/build/BUILD/ldc-1.1.0-beta4-src/build/bin/ldc2+0x10040c7c)
#52 0x00003fff79fc4d88 generic_start_main.isra.0 (/lib64/power8/libc.so.6+0x44d88)
#53 0x00003fff79fc4fb8 __libc_start_main (/lib64/power8/libc.so.6+0x44fb8)
Invalid operand types for ICmp instruction
  %7 = icmp slt float %6, 0.000000e+00
LLVM ERROR: Broken function found, compilation aborted!
make[2]: *** [runtime/CMakeFiles/druntime-ldc.dir/build.make:157: runtime/src/core/math.o] Error 1
```

**Comment by kinke** [comment]
Thx for the report. The only LDC guy having access to PowerPC hardware currently lacks the time to work on LDC, so a fix will take some time.
With the codegen log, we may be able to pinpoint the problem though. So it'd be great if you could compile `core.math` with the self-bootstrapped beta4 using the `-vv` switch log and post the last lines before the crash.

**Comment by kalev** [comment]
Done! https://kojipkgs.fedoraproject.org//work/tasks/1431/16671431/build.log -- I added -vv to the last ldc2 command there.

**Comment by kinke** [comment]
I was on the wrong track here and cannot reproduce the issue with a self-bootstrapped LDC v1.1-beta5 (on Windows) using your triple. It's almost certainly a duplicate of https://github.com/ldc-developers/ldc/issues/1652, and fixing ABI issues without hardware to test on is hard, so we'll have to wait for @redstar.

**Comment by kinke** [comment]
@kalev: It would be awesome if you could give https://github.com/ldc-developers/ldc/pull/1905 a try.

**Comment by kalev** [comment]
Commented in https://github.com/ldc-developers/ldc/pull/1905

**Comment by kinke** [comment]
Fixed by #1905.
